src/main/java/edu/kit/compiler/semantic/DetailedNameTypeAstVisitor.java

```diff
@@ -125,210 +127,252 @@ public DataType visit(ClassNode classNode) {
+            semanticError(returnStatementNode, "method requires a return value of type %s", expectedReturnType.get().getRepresentation(stringTable));
         }
 
-        return VoidType;
+        return Optional.empty();
     }
 
-    public DataType visit(ExpressionStatementNode expressionStatementNode) {
+    public Optional<DataType> visit(ExpressionStatementNode expressionStatementNode) {
         expressionStatementNode.getExpression().accept(this);
 
-        return VoidType;
+        return Optional.empty();
     }
 
-    public DataType visit(BinaryExpressionNode binaryExpressionNode) {
-        DataType leftSideType = binaryExpressionNode.getLeftSide().accept(this);
-        DataType rightSideType = binaryExpressionNode.getRightSide().accept(this);
+    private Optional<DataType> applyResultType(ExpressionNode expressionNode, Optional<DataType> resultType_) {
+        resultType_.ifPresentOrElse(resultType -> {
+            expressionNode.setResultType(resultType);
+        }, () -> {
+            expressionNode.setHasError(true);
+        });
+        return resultType_;
+    }
+
+    private <ReferenceExpressionNode extends ExpressionNode & Reference> void applyDefinition(ReferenceExpressionNode referenceNode, Optional<? extends Definition> definition_) {
+        definition_.ifPresentOrElse(definition -> {
+            referenceNode.setDefinition(definition);
+        }, () -> {
+            referenceNode.setHasError(true);
+        });
+    }
+
+    public Optional<DataType> visit(BinaryExpressionNode binaryExpressionNode) {
+        Optional<DataType> leftSideType_ = binaryExpressionNode.getLeftSide().accept(this);
+        Optional<DataType> rightSideType_ = binaryExpressionNode.getRightSide().accept(this);
 
         BinaryOperator operator = binaryExpressionNode.getOperator();
 
-        DataType resultType;
+        Optional<DataType> resultType_;
         switch (operator) {
             case Assignment:
-                if (!leftSideType.isCompatibleTo(rightSideType)) {
-                    semanticError(binaryExpressionNode, "the two sides of an assignment must have compatible types");
-                }
+                leftSideType_.ifPresent(leftSideType -> rightSideType_.ifPresent(rightSideType -> {
+                    if (!leftSideType.isCompatibleTo(rightSideType)) {
+                        semanticError(binaryExpressionNode, "the two sides of an assignment must have compatible types");
+                    }
+                }));
 
-                resultType = leftSideType;
+                resultType_ = leftSideType_;
                 break;
             case Equal:
             case NotEqual:
-                if (!leftSideType.isCompatibleTo(rightSideType)) {
-                    semanticError(binaryExpressionNode, "the two arguments of %s must have compatible types", operator);
-                }
+                leftSideType_.ifPresent(leftSideType -> rightSideType_.ifPresent(rightSideType -> {
```

Comment:
You use this double `ifPresent` construct quite a few times, you could consider a helper method for this to make it more readable (something like `ifBoth(Optional<S>, Optional<T>, BiFunction<S, T, R>)`).

Reply:
I'm unsure if it would be more readable, but I can change it if it helps.